File: Server/src/atraxi/server/Game.java
package atraxi.server;

import atraxi.core.Player;
import atraxi.core.util.Logger;
import atraxi.core.world.World;
import atraxi.server.networking.ServerUtil;

import java.math.BigDecimal;
import java.util.ArrayList;

public class Game implements Runnable
{
    /**
     * The desired time per frame (in nanoseconds) from which to adjust the speed of all events. 60fps or 1,000,000,000ns/60
     */
    public static final long OPTIMALFRAMETIME = 16666666;
    private static final long MINIMUMFRAMETIME = 8333333;
    private static ArrayList<Player> players;
    private static ArrayList<World> worlds;
    private static ServerUtil serverUtil;
    
    public Game(ArrayList<Player> players, ArrayList<World> worlds)
    {
        Game.players = players;
        Game.worlds = worlds;
        serverUtil = new ServerUtil(players.size());
        new Thread(serverUtil).start();
    }
    
    public static ArrayList<Player> getPlayerList()
    {
        return players;
    }

    public static World getWorld(int index)
    {
        return worlds.get(index);
    }
    
    private void gameLoop(BigDecimal timeAdjustment)
    {
//        for(World world : worlds)
//        {
//            for(Entity entity : world.getEntityList())
//            {
//                entity.doWork(timeAdjustment, paused);
//            }
//        }
    }
    
    @Override
    public void run()
    {
        long beforeTime, actualFrameTime;
        
        beforeTime = System.nanoTime();

        while(true)
        {
            long currentTime = System.nanoTime();
            
            actualFrameTime = currentTime-beforeTime;
            
            if(actualFrameTime > OPTIMALFRAMETIME)
            {
                Logger.log(Logger.LogLevel.warning, new String[] {"Game running behind, skipping " + ((int)((actualFrameTime) / OPTIMALFRAMETIME)) + " frames"});
            }
            if(actualFrameTime < MINIMUMFRAMETIME)
            {
                gameLoop(BigDecimal.valueOf(MINIMUMFRAMETIME).divide(BigDecimal.valueOf(OPTIMALFRAMETIME),8,BigDecimal.ROUND_DOWN));

                beforeTime = currentTime;

                try
                {
                    //Limit framerate to 120fps
                    Thread.sleep((MINIMUMFRAMETIME-actualFrameTime)/1000000);
                }
                catch (InterruptedException e)
                {//TODO: handle this
                    e.printStackTrace();
                }
            }
            else
            {
                gameLoop(BigDecimal.valueOf(actualFrameTime).divide(BigDecimal.valueOf(OPTIMALFRAMETIME),
                                                                    8,
                                                                    BigDecimal.ROUND_DOWN));

                beforeTime = currentTime;
            }
        }
    }

    public static void processTurn()
    {
        players.parallelStream().forEach(Player::processSetStatesThisTurn);
        players.parallelStream().forEach(Player::processBuildsThisTurn);
        players.parallelStream().forEach(Player::processAttacksThisTurn);
        players.parallelStream().forEach(Player::processMovesThisTurn);
    }
}

File: Server/src/atraxi/server/Game.kt
package atraxi.server

import atraxi.core.Player
import atraxi.core.util.Logger
import atraxi.core.world.World
import atraxi.server.networking.ServerUtil

import java.math.BigDecimal
import java.util.ArrayList

class Game(players: ArrayList<Player>, worlds: ArrayList<World>) : Runnable {

    init {
        Game.playerList = players
        Game.worlds = worlds
        serverUtil = ServerUtil(players.size)
        Thread(serverUtil, "ServerSocket").start()
    }

    private fun gameLoop(timeAdjustment: BigDecimal) {
        //        for(World world : worlds)
        //        {
        //            for(Entity entity : world.getEntityList())
        //            {
        //                entity.doWork(timeAdjustment, paused);
        //            }
        //        }
    }

    override fun run() {
        var beforeTime: Long
        var actualFrameTime: Long

        beforeTime = System.nanoTime()

        while (true) {
            val currentTime = System.nanoTime()

            actualFrameTime = currentTime - beforeTime

            if (actualFrameTime > OPTIMALFRAMETIME) {
                Logger.log(Logger.LogLevel.warning, arrayOf("Game running behind, skipping " + (actualFrameTime / OPTIMALFRAMETIME).toInt() + " frames"))
            }
            if (actualFrameTime < MINIMUMFRAMETIME) {
                gameLoop(BigDecimal.valueOf(MINIMUMFRAMETIME).divide(BigDecimal.valueOf(OPTIMALFRAMETIME), 8, BigDecimal.ROUND_DOWN))

                beforeTime = currentTime

                try {
                    //Limit framerate to 120fps
                    Thread.sleep((MINIMUMFRAMETIME - actualFrameTime) / 1000000)
                } catch (e: InterruptedException) {//TODO: handle this
                    e.printStackTrace()
                }

            } else {
                gameLoop(BigDecimal.valueOf(actualFrameTime).divide(BigDecimal.valueOf(OPTIMALFRAMETIME),
                        8,
                        BigDecimal.ROUND_DOWN))

                beforeTime = currentTime
            }
        }
    }

    companion object {
        /**
         * The desired time per frame (in nanoseconds) from which to adjust the speed of all events. 60fps or 1,000,000,000ns/60
         */
        val OPTIMALFRAMETIME: Long = 16666666
        private val MINIMUMFRAMETIME: Long = 8333333
        var playerList: ArrayList<Player>? = null
            private set
        private var worlds: ArrayList<World>? = null
        private var serverUtil: ServerUtil

        fun getWorld(index: Int): World {
            return worlds!![index]
        }

        val worldCount: Int
            get() = worlds!!.size

        fun processTurn() {
            Logger.log(Logger.LogLevel.debug, arrayOf("End turn received from all players, processing all actions."))
            playerList!!.parallelStream().forEach(Consumer<Player> { it.processSetStatesThisTurn() })
            playerList!!.parallelStream().forEach(Consumer<Player> { it.processBuildsThisTurn() })
            playerList!!.parallelStream().forEach(Consumer<Player> { it.processAttacksThisTurn() })
            playerList!!.parallelStream().forEach(Consumer<Player> { it.processMovesThisTurn() })
        }
    }
}